### PR TITLE
Add PostgresqlCheck actor

### PR DESCRIPTION
**Pull request by panovotn**
This actor collects information about Red Hat Signed PostgreSQL
installation and warns user about possible manual intervention
requirements where applicable.

### CODE REVIEW

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                       'version': version,
+                       'schema': schema,
+                       'description': description})
+
+    return result
+
+def generate_report(packages):
+    """ Generate reports informing user about possible manual intervention requirements """
+
+    DOC_URL = 'https://www.postgresql.org/docs/10/upgrading.html'
```
Comment (hhorak) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: We should ideally link to internal upgrade steps (currently available only in 8.0 Beta docs), and we don't know a correct link that will be ready after GA. How to handle this? Any idea, @pirat89 @bocekm?

Comment (vinzenz) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: @hhorak just post a patch once you have it, or have an RedHat KB article you can with some text and update the link in that article once you have the right link

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    for rpm_pkgs in api.consume(InstalledRedHatSignedRPM):
+        return [pkg for pkg in rpm_pkgs.items if pkg.name.startswith('postgresql')]
+
+def get_postgresql_extensions():
+    """ Get list of PostgreSQL extensions """
+    result = []
+    try:
+        output = run([
+            '/usr/bin/su',
+            '-l', 'postgresql'
```
Comment (hhorak) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: This seems to be an issue, the superuser is called `postgres`. Is it working this way?

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    MISSING_EXTENSIONS = ['dummy_seclabel', 'test_parser', 'tsearch2']
+
+    extensions = get_postgresql_extensions()
+
+    for pkg in packages:
+        # Show documentation url if postgresql-server installed
+        if pkg.name == 'postgresql-server':
+            title = 'PostgreSQL server installation detected'
+            severity = 'high'
+            summary = 'PostgreSQL server will be upgraded. Additional steps might be required.\n\n' \
```
Comment (hhorak) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: I'd be be more concrete here. For example:
"PostgreSQL server component will be upgraded. Since RHEL-8 includes PostgreSQL server 10 by default, which is incompatible with 9.2 included in RHEL-7, it is necessary to proceed with additional steps for complete upgrade of the PostgreSQL data. Back-up your data before proceeding with the upgrade and read more about upgrade steps at <KB_article_url>."

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    result = []
+    try:
+        output = run([
+            '/usr/bin/su',
+            '-l', 'postgresql'
+            '-c', 'psql -tAc \'\dx\''
+        ], split=True)['stdout']
+    except CalledProcessError:
+        api.current_logger().debug('Unable to list PostgreSQL extensions: psql failed to run')
+        return False
```
Comment (hhorak) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Wouldn't False here cause error later, when the output of this function is taken as array? Maybe returning an empty array in this case would be better.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            summary = 'PostgreSQL server will be upgraded. Additional steps might be required.\n\n' \
+                      'Read more: {}'.format(DOC_URL)
+
+            reporting.report_generic(title=title, severity=severity, summary=summary)
+
+        # Show warning about extensions removed from postgresql-contrib
+        if pkg.name == 'postgresql-contrib' and len(extensions):
+            extensions_removed = [ext for ext in extensions if ext.name in MISSING_EXTENSIONS]
+
+            if len(extensions_removed):
```
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]:
```suggestion
            if extensions_removed:
```

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        if pkg.name == 'postgresql-server':
+            title = 'PostgreSQL server installation detected'
+            severity = 'high'
+            summary = 'PostgreSQL server will be upgraded. Additional steps might be required.\n\n' \
+                      'Read more: {}'.format(DOC_URL)
+
+            reporting.report_generic(title=title, severity=severity, summary=summary)
+
+        # Show warning about extensions removed from postgresql-contrib
+        if pkg.name == 'postgresql-contrib' and len(extensions):
```
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]:
```suggestion
        if pkg.name == 'postgresql-contrib' and extensions:
```

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 4 above)
Comment (vinzenz) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Its probably even better to return None as it’s not an empty list of extensions but the command failed

If you check if value instead of if len(value) then this will work with false none and an empty list

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 2 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Definitely isn't. This happened while renaming get_postgresql_packages function. Thanks for pointing it out. Fixed.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 5 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 6 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed as well.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 4 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: I agree. Returning None seems much more universal as @vinzenz pointed out. Fixed.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 3 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: I've used text you've provided here (thank you), and included a KB article URL. That article is empty ATM, but will be filled in as soon as I get proper permissions. This should be sufficient for the sake of merging this PR.

**Comment by UncleAlbie** [comment]
As I've addressed all comments, can we merge?

**Comment by pirat89** [comment]
@UncleAlbie If I see right, there are still unresolved discussions - between those,at least one has not been adressed.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 1 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: As I've mentioned in another conversation, the URL is now changed to the KB article. This should be fixed as well.

**Comment by UncleAlbie** [comment]
@pirat89 Those have been fixed as well, I just mentioned those fixes only in one of the conversations. My mistake. Now it should be ready.

**Comment by hhorak** [comment]
@pirat89 just checking whether there is something else that needs to be done before merging. @UncleAlbie claims the requested changes were addressed.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        return [pkg for pkg in rpm_pkgs.items if pkg.name.startswith('postgresql')]
+
+def get_postgresql_extensions():
+    """ Get list of PostgreSQL extensions """
+    result = []
+    try:
+        output = run([
+            '/usr/bin/su',
+            '-l', 'postgres'
+            '-c', 'psql -tAc \'\dx\''
```
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: This seems to be an error:
```suggestion
            '-c', 'psql -tAc \'\\dx\''
```

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                      'server 10 by default, which is incompatible with 9.2 included in RHEL-7, it is\n' \
+                      'necessary to proceed with additional steps for complete upgrade of the\n' \
+                      'PostgreSQL data. Back-up your data before proceeding with the upgrade and read\n' \
+                      'more about upgrade steps at {}.'.format(DOC_URL)
+
+            reporting.report_generic(title=title, severity=severity, summary=summary)
+
+        # Show warning about extensions removed from postgresql-contrib
+        if pkg.name == 'postgresql-contrib' and extensions:
+            extensions_removed = [ext for ext in extensions if ext.name in MISSING_EXTENSIONS]
```
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]:
```suggestion
            extensions_removed = [ext for ext in extensions if ext['name'] in MISSING_EXTENSIONS]
```

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                title = 'PostgreSQL extensions removed from contrib detected'
+                severity = 'high'
+                summary = 'Following installed extensions won\'t be part of PostgreSQL distribution' \
+                          'after upgrade:\n{}'.format(extensions_removed_newline)
+
+                reporting.report_generic(title=title, severity=severity, summary=summary)
+
+    # Show list of installed extensions not shipped with postgresql-contrib
+    # and warn about possible incompatibility of manually installed extensions
+    extensions_public = [ext for ext in extensions if ext.schema == 'public']
```
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]:
```suggestion
    extensions_public = [ext for ext in extensions if ext['schema'] == 'public']
```

Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: And out of curiosity, who says that the extensions are installed to public schema? Wouldn't it be possible to have it installed to a different schema? I mean I know it's common to do it in the public schema but I am not sure if someone might use a different schema than the public schema, since the CREATE EXTENSION has a schema parameter.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    """ Get list of PostgreSQL extensions """
+    result = []
+    try:
+        output = run([
+            '/usr/bin/su',
+            '-l', 'postgres'
+            '-c', 'psql -tAc \'\dx\''
+        ], split=True)['stdout']
+    except CalledProcessError:
+        api.current_logger().debug('Unable to list PostgreSQL extensions: psql failed to run')
```
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Are you sure that should be a debug message?

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
+from leapp.models import InstalledRedHatSignedRPM
+from leapp.libraries.common import reporting
+from leapp.libraries.stdlib import CalledProcessError, api, run
+
+def get_postgresql_packages():
+    """ Get list of installed PostgreSQL packages """
+
+    for rpm_pkgs in api.consume(InstalledRedHatSignedRPM):
+        return [pkg for pkg in rpm_pkgs.items if pkg.name.startswith('postgresql')]
```
[the remark this answers is not in the document]
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+def generate_report(packages):
+    """ Generate reports informing user about possible manual intervention requirements """
+
+    DOC_URL = 'https://access.redhat.com/articles/4055641'
+    MISSING_EXTENSIONS = ['dummy_seclabel', 'test_parser', 'tsearch2']
+
+    extensions = get_postgresql_extensions()
+
+    for pkg in packages:
```
[the remark this answers is not in the document]
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        # Show documentation url if postgresql-server installed
+        if pkg.name == 'postgresql-server':
```
[the remark this answers is not in the document]
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+            summary = 'PostgreSQL server component will be upgraded. Since RHEL-8 includes PostgreSQL\n' \
+                      'server 10 by default, which is incompatible with 9.2 included in RHEL-7, it is\n' \
+                      'necessary to proceed with additional steps for complete upgrade of the\n' \
+                      'PostgreSQL data. Back-up your data before proceeding with the upgrade and read\n' \
+                      'more about upgrade steps at {}.'.format(DOC_URL)
+
+            reporting.report_generic(title=title, severity=severity, summary=summary)
+
+        # Show warning about extensions removed from postgresql-contrib
+        if pkg.name == 'postgresql-contrib' and extensions:
```
[the remark this answers is not in the document]
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 7 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 8 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 9 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Suggested change pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 10 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Changed to 'warning' and pushed as fixup.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 9 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: You're right. However, I'd suggest solving this in another PR as the public schema really is commonly used.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+            title = 'PostgreSQL extensions removed from contrib detected'
+            severity = 'high'
+            summary = 'Following installed extensions won\'t be part of PostgreSQL distribution' \
+                      'after upgrade:\n{}'.format(extensions_removed_newline)
+
+            reporting.report_generic(title=title, severity=severity, summary=summary)
+
+    # Show list of installed extensions not shipped with postgresql-contrib
+    # and warn about possible incompatibility of manually installed extensions
+    extensions_public = [ext for ext in extensions if ext['schema'] == 'public']
```
Comment (vinzenz) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Needs check for ‘extensions’ be Ayse it can be None

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    if 'postgresql-server' in packages:
+        title = 'PostgreSQL server installation detected'
+        severity = 'high'
+        summary = 'PostgreSQL server component will be upgraded. Since RHEL-8 includes PostgreSQL\n' \
+                  'server 10 by default, which is incompatible with 9.2 included in RHEL-7, it is\n' \
+                  'necessary to proceed with additional steps for complete upgrade of the\n' \
+                  'PostgreSQL data. Back-up your data before proceeding with the upgrade and read\n' \
+                  'more about upgrade steps at {}.'.format(DOC_URL)
+
+        reporting.report_generic(title=title, severity=severity, summary=summary)
```
Comment (shaded-enmity) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Please use `report_with_links` if you want to include links - https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/el7toel8/libraries/reporting.py#L40

**Comment by pirat89** [comment]
@UncleAlbie Please, do not forget to rebase against the current master when you resolve the additional required changes.

**Review by vinzenz — DISMISSED** [review]
Please apply the comments from Pavel and me and once more rebase against current master. Thanks

**Comment by vinzenz** [comment]
Ping? Please address comments and rebase, otherwise this PR is almost ready.

**Comment by UncleAlbie** [comment]
I just need to finalize tests for the actor. Hopefully, I'll get to it by the end of this week.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 15 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed. Thanks for pointing that out.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py (hunk 16 above)
Comment (UncleAlbie) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: I've changed the report to `report_with_links`.

**Comment by hhorak** [comment]
There is also the article that needs to be published before this PR is merged: https://access.redhat.com/articles/4055641

**Comment by UncleAlbie** [comment]
Tests for this actor is now failing on "reporting" module import .. could this be a LEAPP issue, or am I missing something?

**Comment by bocekm** [comment]
We have changed the reporting feature in the leapp framework. Please take a look at this PR on examples how to update reporting function calls:
https://github.com/oamg/leapp-repository/pull/295/files
Here's a documentation on how to use the new reporting:
https://leapp.readthedocs.io/en/latest/actor-rhel7-to-rhel8.html#producing-data-for-other-actors-and-reporting

**Comment by bocekm** [comment]
leapp-ci build

**Comment by bocekm** [comment]
@fellipeh, @asilveir, you can take a look at this PR if you have time. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+from leapp import reporting
+from leapp.reporting import create_report
+from leapp.libraries.stdlib import CalledProcessError, api, run
+
+
+def get_postgresql_packages():
+    """ Get list of installed PostgreSQL packages """
+
+    for rpm_pkgs in api.consume(InstalledRedHatSignedRPM):
+        return {pkg.name for pkg in rpm_pkgs.items if pkg.name.startswith('postgresql')}
```
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Currently, only RPMs that are checked in the actor are `postgresql-server` and `postgresql-contrib`. From that point it doesn't seem necessary to use this function. Instead of that, wouldn't be better to use `has_package` function, that will provide that info? See:

```
   from leapp.libraries.common.rpms import has_package

   if has_package(InstalledRedHatSignedRPM, 'postgresql-server'):
```

We prefer this library function to be used instead, but it is not blocker. The code looks good anyway, but I haven't tested it..

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+
+class create_report_mocked(object):
+    def __init__(self):
+        self.called = 0
+        self.report_fields = None
+
+    def __call__(self, **report_fields):
+        self.called += 1
+        self.report_fields = report_fields
```
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py]: The `produce_mocked` function is not ideal to test produced reports and it is duplicit code. Similar to `create_report_mocked`, which is not completely same, but pretty similar. Look at `create_report_mocked` in `libraries/testutils.py` instead. There are already some actors using it. And you can remove the `produce_mocked` completely as `produce` is part of the `create_report ` function.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def test_actor_execution_without_unsigned_data(monkeypatch):
+    def consume_postgresql_packages_mocked(*models):
+        installed_rpm = []
+        yield InstalledSignedRPM(items=installed_rpm)
+    monkeypatch.setattr(api, "consume", consume_postgresql_packages_mocked)
+    monkeypatch.setattr(api, "produce", produce_mocked())
```
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py]:
```suggestion
```
This is not needed here.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+
+def test_actor_execution_with_postgresql(monkeypatch):
+    def consume_postgresql_packages_mocked(*models):
+        installed_rpm = [
+            RPM(name='postgresql', version='0.1', release='1.sm01', epoch='1', packager=RH_PACKAGER, arch='noarch',
+                pgpsig='SOME_OTHER_SIG_X')]
+        yield InstalledSignedRPM(items=installed_rpm)
+
+    monkeypatch.setattr(api, "consume", consume_postgresql_packages_mocked)
+    monkeypatch.setattr(api, "produce", produce_mocked())
```
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py]:
```suggestion
    monkeypatch.setattr(api, "produce", produce_mocked())
```

Same here.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+def test_actor_execution_without_unsigned_data(monkeypatch):
+    def consume_postgresql_packages_mocked(*models):
+        installed_rpm = []
+        yield InstalledSignedRPM(items=installed_rpm)
+    monkeypatch.setattr(api, "consume", consume_postgresql_packages_mocked)
+    monkeypatch.setattr(api, "produce", produce_mocked())
+    monkeypatch.setattr(api, "show_message", lambda x: True)
+    monkeypatch.setattr(reporting, "create_report", create_report_mocked())
+
+    packages = library.get_postgresql_packages()
```
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/postgresqlcheck/tests/test_postgresqlcheck.py]: I prefer to create the `packages` input directly (if it is valid after possible changes in the actor I mentioned earlier). It doesn't make so much sense to test it in this way. If you want to keep the get_postgresql_packages function, test it separately please.

**Comment by UncleAlbie** [comment]
I've rewritten the actor from the ground up. There's been some changes in the framework and it just made sense to me. It's much simpler now.

As all checks have passed, it should be ready for review. However, please, don't merge yet. There's still some work around the KB article needed. Thanks!

**Comment by pirat89** [comment]
@UncleAlbie thanks for info. So we will do review without merging just to give you feedback.

**Comment by shaded-enmity** [comment]
@oamg/developers @UncleAlbie Are there any blockers? The code looks very good to me.

**Comment by bocekm** [comment]
@UncleAlbie, is this PR still in progress? Is it ready for another round of reviews?

**Comment by bocekm** [comment]
@UncleAlbie, ping :)

**Comment by Rezney** [comment]
@UncleAlbie is there anything we can do to move this forward? Thanks...

**Comment by UncleAlbie** [comment]
This PR should be ready for (review and) merge.

**Review by pirat89 — CHANGES_REQUESTED** [review]
@UncleAlbie I apologize for really long waiting for review. I've just discovered couple of things now:
- the actors library needs to be renamed to ensure the name is uniq - it's to ensure the unit-tests can be executed correctly in faster way, instructions about naming of test files and libraries are here:
- https://leapp.readthedocs.io/en/latest/unit-testing.html#naming-conventions
- https://leapp.readthedocs.io/en/latest/unit-testing.html#testing-private-actor-library-unit-tests

In short, call actor's library e.g. "postgresqlcheck.py" and fix imports. 

Other stuff in comments/suggestions below. I think that the rest seems fine after the stuff will be fixed. I will give it another look tomorrow during proper work-time.

**Comment by pirat89** [comment]
Note: the article is not published yet.

**Comment by Rezney** [comment]
@pirat89 is there anything we can do to move this forward?

**Comment by pirat89** [comment]
@Rezney  the referred article is not published yet (https://access.redhat.com/articles/4055641), probably some additional refactoring is needed and several required changes have not been addressed. We can sync with guys to push it, but currently there is not so much time. Probably we could sync with @hhorak (ping, wdyt Honzo?) about this PR to decide what to do with it (finish it? close it?). In case of the finalisation work, we should schedule a timeplan to ensure that someone from our team and someone from database team have time to finish this. As well we should sync about the testing.

**Comment by hhorak** [comment]
We agreed with @pirat89 and @panovotn to finish this in 1st part of June, the article should be published till then.
Also tracking this change in BZ https://bugzilla.redhat.com/show_bug.cgi?id=1962081 in order to make syncing with QE easier.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+# Title for postgresql-server report
+report_server_inst_title = 'PostgreSQL (postgresql-server) has been detected on your system'
+# Summary for postgresql-server report
+report_server_inst_summary = 'PostgreSQL server component will be upgraded. Since ' \
+                             'RHEL-8 includes PostgreSQL server 10 by default, ' \
+                             'which is incompatible with 9.2 included in RHEL-7, ' \
+                             'it is necessary to proceed with additional steps ' \
+                             'for complete upgrade of the PostgreSQL data. Back-up ' \
+                             'your data before proceeding with the upgrade and read ' \
+                             'more about upgrade steps in the knowledge base article.'
```
[the remark this answers is not in the document]
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed according to your suggestion. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+
+# List of dropped extensions from postgresql-contrib package
+report_contrib_inst_dropext = ['dummy_seclabel', 'test_parser', 'tsearch2']
+
+# Title for postgresql-contrib report
+report_contrib_inst_title = 'PostgreSQL (postgresql-contrib) has been detected on your system'
+# Summary for postgresql-contrib report
+report_contrib_inst_summary = 'Please note that some extensions have been dropped from the ' \
+                              'postgresql-contrib package and might not be available after upgrade:\n' \
+                              '{}'.format('\n'.join(report_contrib_inst_dropext))
```
[the remark this answers is not in the document]
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed according to your suggestion. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+    redirect them to online documentation for the upgrade process.
+
+    """
+    reporting.create_report([
+        reporting.Title(report_server_inst_title),
+        reporting.Summary(report_server_inst_summary),
+        reporting.Severity(reporting.Severity.HIGH),
+        reporting.Tags([reporting.Tags.SERVICES]),
+        reporting.ExternalLink(title=report_server_inst_link_title,
+                               url=report_server_inst_link_url)
```
[the remark this answers is not in the document]
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed according to your suggestion. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+    Should remind user about present PostgreSQL contrib package
+    installation and provide them with a list of extensions no longer
+    shipped with this package.
+
+    """
+    reporting.create_report([
+        reporting.Title(report_contrib_inst_title),
+        reporting.Summary(report_contrib_inst_summary),
+        reporting.Severity(reporting.Severity.HIGH),
+        reporting.Tags([reporting.Tags.SERVICES])
```
[the remark this answers is not in the document]
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: Fixed according to your suggestion. Thanks.

File: repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+                             'which is incompatible with 9.2 included in RHEL-7, ' \
+                             'it is necessary to proceed with additional steps ' \
+                             'for complete upgrade of the PostgreSQL data. Back-up ' \
+                             'your data before proceeding with the upgrade and read ' \
+                             'more about upgrade steps in the knowledge base article.'
+
+# Link text for postgresql-server report
+report_server_inst_link_title = 'Learn more in this KB article.'
+# Link URL for postgresql-server report
+report_server_inst_link_url = 'https://access.redhat.com/articles/4055641'
```
[the remark this answers is not in the document]
Comment (panovotn) [reply repos/system_upgrade/el7toel8/actors/postgresqlcheck/libraries/library.py]: The article content is actually duplicate to the upgrade section in our PostgreSQL documentation. Therefore, I've changed the URL to a link to that section.

IMO, separate article isn't needed for this version of the actor.

**Review by pirat89 — APPROVED** [review]
Lgtm. Regarding tests it seems it ok. I will test it manually yet.

**Comment by pirat89** [comment]
leapp-ci build

**Comment by pirat89** [comment]
leapp-ci build

**Comment by pirat89** [comment]
:snail: :turtle:  After two years and two months: :fireworks: :confetti_ball:  :partying_face:  :champagne:

**Comment by pirat89** [comment]
Btw, just for the completness, here is part of the report when postgresql-server and postgresql-contrib packages are installed:
```
----------------------------------------
Risk Factor: medium
Title: PostgreSQL (postgresql-server) has been detected on your system
Summary: PostgreSQL server component will be upgraded. Since RHEL-8 includes PostgreSQL server 10 by default, which is incompatible with 9.2 included in RHEL-7, it is necessary to proceed with additional steps for the complete upgrade of the PostgreSQL data.
Remediation: [hint] Back up your data before proceeding with the upgrade and follow steps in the documentation section "Migrating to a RHEL 8 version of PostgreSQL" after the upgrade.
Key: 6966486bbbf49710d1d9f4e4f6b612217aca38a1
----------------------------------------
Risk Factor: medium
Title: PostgreSQL (postgresql-contrib) has been detected on your system
Summary: Please note that some extensions have been dropped from the postgresql-contrib package and might not be available after the upgrade:
    - dummy_seclabel
    - test_parser
    - tsearch2
Key: e0703e8e14b8c2eee885de3fb451c9d072b89ed6
```

I've tested even the case when just the postgresql-server package is installed, so just the one report is generated as expected.
